### PR TITLE
Signposting api

### PR DESCRIPTION
**Pull request by elena-vi**
**What**  
allow conversations to be sent and sign posts to be saved
**Why**  
{why are we making this change?}

**Anything else?**  
 - Have you added any new third-party libraries?
 - Are any new environment variables or configuration values needed?
 - Anything else in this PR that isn't covered by the what/why?

### CODE REVIEW

File: lib/domain/conversation.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+//     name: userName,
+//     organisation: userOrganisation,
+//     email: userEmail
+//   },
+//   discussedServices: discussedServices,
+//   signPostingMessage: signPostingMessage,
+//   id: undefined,
+//   systemIds: undefined,
+//   created: expect.any(String)
+// });
```
Comment (kosiakkatrina) [inline lib/domain/conversation.js:71]: does it need to be commented?

File: pages/api/conversations/index.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+//   },
+//   {
+//     id: 2,
+//     name: 'Chocolate Deliveries',
+//     contactEmail: 'contact@chocolatedeliveries.com',
+//     referralEmail: 'referral@chocolatedeliveries.com',
+//     website: 'www.chocolatedeliveries.com',
+//     address: '123 Other st, E3 1P'
+//   }
+// ]
```
Comment (kosiakkatrina) [inline pages/api/conversations/index.js:94]: 👀

**Review by kosiakkatrina — APPROVED** [review]

File: lib/use-cases/create-conversation.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      systemIds,
+      emails: {
+        conversationEmailId: signPostingEmail.response.data?.id
+      }
+    });
+
+    return {
+      id: conversation.id,
+      errors: errors
+    };
```
Comment (Duslerke) [inline lib/use-cases/create-conversation.js:37]: Maybe we should add error handling for writing `conversation` to database as well?

Like add another:
```
if (conversation.response.status > 201) {
      errors.push('Failed to write the record into the database.');
}
```

or handle it some other way at Ln33.

File: lib/use-cases/create-conversation.test.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      phone,
+      referralReason,
+      conversationNotes,
+      userName,
+      userOrganisation,
+      userEmail
+    }))
+  };
+  const sendSignpostEmail = {
+    execute: jest.fn(() => ({ response: { status: 300 } }))
```
Comment (Duslerke) [inline lib/use-cases/create-conversation.test.js:58]: What magic is this? 😅 Shouldn't this fail due to errors being returned due to this code block in the `create-conversation` use-case (file above): 

```
if (signPostingEmail.response.status > 201) {
      errors.push('Failed to send sign posting email to resident');
}
```

Comment (Duslerke) [reply lib/use-cases/create-conversation.test.js:58]: Oh, never mind... I see that the test we have is actually 2 distinct tests & we're not asserting for errors array to be zero length... it's sort of like both happy & bad path at the same time 😁. Would suggest to separate those out, but I assume we're out of time to do refactoring like that.

File: lib/use-cases/send-signpost-email.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+export default class SendSignpostEmail {
```
Comment (Duslerke) [inline lib/use-cases/send-signpost-email.js:1]: This is where I probably don't understand Clean Architecture well enough...
image


Is use-case having a use-case within its dependencies valid? Shouldn't it be one of the two?:
1. Controller (endpoint) calling 2 use-cases (create-converstation & send-signpost)
2. `CreateConverstation` use-case calling 2 gateways (create-converstation & notify)

File: pages/api/conversations/index.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        {
+          name: 'lastName',
+          failureMessage: 'last name is required',
+          validate: ({ body }) => body.lastName && body.lastName.length > 0
+        },
+        {
+          name: 'email',
+          failureMessage: 'resident email is required',
+          validate: ({ body }) => body.email && body.email.length > 0
+        },
```
Comment (Duslerke) [inline pages/api/conversations/index.js:25]: Ln23 failureMessage specifies that email field belongs to resident. Should we do the same for firstName & lastName?

File: pages/api/conversations/index.test.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        lastName: 'taylor',
+        userEmail: 'user@email.com',
+        discussedServices: [{}],
+        signPostingMessage: 'msg',
+        email: 'resident@email.com'
+      }
+    });
+    expect(response.statusCode).toBe(500);
+  });
+});
```
Comment (Duslerke) [inline pages/api/conversations/index.test.js:131]: Awesome! We have quite a number of bad-path tests here. I was wondering whether we should include the 403, or is it being returned by default (so no need to test it)?

**Review by Duslerke — APPROVED** [review]
Looks good.

Mentioned couple things and questions here & there, I doubt they're a priority to address right now, so... I don't mind if they don't get addressed if we don't have the time for it.

**Pull request merged**